batch/batch/front_end/query/query_v1.py

```diff
@@ -85,47 +87,110 @@ def parse_list_batches_query_v1(user: str, q: str, last_batch_id: Optional[int])
 
     sql = f"""
 WITH base_t AS (
-  SELECT batches.*,
+  SELECT batches.*, job_groups.batch_id, job_groups.job_group_id,
     job_groups_cancelled.id IS NOT NULL AS cancelled,
     job_groups_n_jobs_in_complete_states.n_completed,
     job_groups_n_jobs_in_complete_states.n_succeeded,
     job_groups_n_jobs_in_complete_states.n_failed,
     job_groups_n_jobs_in_complete_states.n_cancelled
-  FROM batches
+  FROM job_groups
+  LEFT JOIN batches ON batches.id = job_groups.batch_id
   LEFT JOIN billing_projects ON batches.billing_project = billing_projects.name
   LEFT JOIN job_groups_n_jobs_in_complete_states
-    ON batches.id = job_groups_n_jobs_in_complete_states.id
+    ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
   LEFT JOIN job_groups_cancelled
-    ON batches.id = job_groups_cancelled.id
+    ON job_groups.batch_id = job_groups_cancelled.id AND job_groups.job_group_id = job_groups_cancelled.job_group_id
   STRAIGHT_JOIN billing_project_users ON batches.billing_project = billing_project_users.billing_project
   WHERE {' AND '.join(where_conditions)}
-  ORDER BY id DESC
+  ORDER BY batch_id DESC
   LIMIT 51
 )
 SELECT base_t.*, cost_t.cost, cost_t.cost_breakdown
 FROM base_t
 LEFT JOIN LATERAL (
   SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
   FROM (
-    SELECT batch_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+    SELECT batch_id, job_group_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
     FROM aggregated_job_group_resources_v3
-    WHERE base_t.id = aggregated_job_group_resources_v3.batch_id
-    GROUP BY batch_id, resource_id
+    WHERE base_t.id = aggregated_job_group_resources_v3.batch_id AND base_t.job_group_id = aggregated_job_group_resources_v3.job_group_id
+    GROUP BY batch_id, job_group_id, resource_id
   ) AS usage_t
   LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
-  GROUP BY batch_id
+  GROUP BY batch_id, job_group_id
 ) AS cost_t ON TRUE
-ORDER BY id DESC;
+ORDER BY batch_id DESC;
 """
 
     return (sql, where_args)
 
 
-def parse_batch_jobs_query_v1(batch_id: int, q: str, last_job_id: Optional[int]) -> Tuple[str, List[Any]]:
+def parse_list_job_groups_query_v1(
+    batch_id: int, job_group_id: int, last_child_job_group_id: Optional[int]
+) -> Tuple[str, List[Any]]:
+    where_conds = [
+        '(job_groups.batch_id = %s)',
+        '(NOT deleted)',
+        '(job_group_self_and_ancestors.ancestor_id = %s AND job_group_self_and_ancestors.level = 1)',
+    ]
+    sql_args = [batch_id, job_group_id]
+
+    if last_child_job_group_id is not None:
+        where_conds.append('(job_groups.job_group_id > %s)')
+        sql_args.append(last_child_job_group_id)
+
+    sql = f"""
+SELECT job_groups.*,
```

Comment:
```
mysql> EXPLAIN SELECT job_groups.*,
    ->   job_groups_cancelled.id IS NOT NULL AS cancelled,
    ->   job_groups_n_jobs_in_complete_states.n_completed,
    ->   job_groups_n_jobs_in_complete_states.n_succeeded,
    ->   job_groups_n_jobs_in_complete_states.n_failed,
    ->   job_groups_n_jobs_in_complete_states.n_cancelled,
    ->   cost_t.cost, cost_t.cost_breakdown
    -> FROM job_groups
    -> LEFT JOIN batches ON batches.id = job_groups.batch_id
    -> LEFT JOIN job_group_self_and_ancestors ON job_group_self_and_ancestors.batch_id = job_groups.batch_id AND job_group_self_and_ancestors.job_group_id = job_groups.job_group_id 
    -> LEFT JOIN job_groups_n_jobs_in_complete_states
    ->    ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
    -> LEFT JOIN job_groups_cancelled
    ->    ON job_groups.batch_id = job_groups_cancelled.id AND job_groups.job_group_id = job_groups_cancelled.job_group_id
    -> LEFT JOIN LATERAL (
    -> SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    -> FROM (
    -> SELECT batch_id, job_group_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    -> FROM aggregated_job_group_resources_v3
    -> WHERE job_groups.batch_id = aggregated_job_group_resources_v3.batch_id AND job_groups.job_group_id = aggregated_job_group_resources_v3.job_group_id
    -> GROUP BY batch_id, job_group_id, resource_id
    -> ) AS usage_t
    -> LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> GROUP BY batch_id, job_group_id
    -> ) AS cost_t ON TRUE
    -> WHERE (job_groups.batch_id = 1110000) AND (NOT deleted) AND (job_group_self_and_ancestors.ancestor_id = 0 AND job_group_self_and_ancestors.level = 1)
    -> ORDER BY job_group_id ASC;
+----+-------------------+--------------------------------------+------------+--------+--------------------------------------------------------------------------------------------------------+-------------+---------+-------------------------------------------+------+----------+------------------------------+
| id | select_type       | table                                | partitions | type   | possible_keys                                                                                          | key         | key_len | ref                                       | rows | filtered | Extra                        |
+----+-------------------+--------------------------------------+------------+--------+--------------------------------------------------------------------------------------------------------+-------------+---------+-------------------------------------------+------+----------+------------------------------+
|  1 | PRIMARY           | batches                              | NULL       | const  | PRIMARY,batches_deleted                                                                                | PRIMARY     | 8       | const                                     |    1 |   100.00 | NULL                         |
|  1 | PRIMARY           | job_groups                           | NULL       | ref    | PRIMARY,job_groups_state_callback,job_groups_time_created,job_groups_time_completed                    | PRIMARY     | 8       | const                                     |    1 |   100.00 | Rematerialize (<derived2>)   |
|  1 | PRIMARY           | job_group_self_and_ancestors         | NULL       | eq_ref | PRIMARY,job_group_self_and_ancestors_ancestor_id_level,job_group_self_and_ancestors_job_group_id_level | PRIMARY     | 16      | const,batch.job_groups.job_group_id,const |    1 |    10.00 | Using where                  |
|  1 | PRIMARY           | job_groups_n_jobs_in_complete_states | NULL       | const  | PRIMARY                                                                                                | PRIMARY     | 8       | const                                     |    1 |   100.00 | Using where                  |
|  1 | PRIMARY           | job_groups_cancelled                 | NULL       | const  | PRIMARY                                                                                                | PRIMARY     | 8       | const                                     |    1 |   100.00 | Using where                  |
|  1 | PRIMARY           | <derived2>                           | NULL       | ALL    | NULL                                                                                                   | NULL        | NULL    | NULL                                      |    2 |   100.00 | Using where                  |
|  2 | DEPENDENT DERIVED | <derived3>                           | NULL       | ALL    | NULL                                                                                                   | NULL        | NULL    | NULL                                      |    2 |   100.00 | Using filesort               |
|  2 | DEPENDENT DERIVED | resources                            | NULL       | eq_ref | resource_id                                                                                            | resource_id | 4       | usage_t.resource_id                       |    1 |   100.00 | NULL                         |
|  3 | DEPENDENT DERIVED | aggregated_job_group_resources_v3    | NULL       | ref    | PRIMARY                                                                                                | PRIMARY     | 8       | batch.job_groups.batch_id                 |   21 |    10.00 | Using where; Using temporary |
+----+-------------------+--------------------------------------+------------+--------+--------------------------------------------------------------------------------------------------------+-------------+---------+-------------------------------------------+------+----------+------------------------------+
9 rows in set, 3 warnings (0.02 sec)
```

Reply:
Here's your proposed change. Let me know which you prefer.

```
mysql> EXPLAIN SELECT job_groups.*,
    ->   job_groups_cancelled.id IS NOT NULL AS cancelled,
    ->   job_groups_n_jobs_in_complete_states.n_completed,
    ->   job_groups_n_jobs_in_complete_states.n_succeeded,
    ->   job_groups_n_jobs_in_complete_states.n_failed,
    ->   job_groups_n_jobs_in_complete_states.n_cancelled,
    ->   cost_t.cost, cost_t.cost_breakdown
    -> FROM job_group_self_and_ancestors
    -> LEFT JOIN job_groups ON job_group_self_and_ancestors.batch_id = job_groups.batch_id AND job_group_self_and_ancestors.job_group_id = job_groups.job_group_id
    -> LEFT JOIN batches ON batches.id = job_groups.batch_id
    -> LEFT JOIN job_groups_n_jobs_in_complete_states
    ->    ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
    -> LEFT JOIN job_groups_cancelled
    ->    ON job_groups.batch_id = job_groups_cancelled.id AND job_groups.job_group_id = job_groups_cancelled.job_group_id
    -> LEFT JOIN LATERAL (
    -> SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    -> FROM (
    -> SELECT batch_id, job_group_id, resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    -> FROM aggregated_job_group_resources_v3
    -> WHERE job_groups.batch_id = aggregated_job_group_resources_v3.batch_id AND job_groups.job_group_id = aggregated_job_group_resources_v3.job_group_id
    -> GROUP BY batch_id, job_group_id, resource_id
    -> ) AS usage_t
    -> LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> GROUP BY batch_id, job_group_id
    -> ) AS cost_t ON TRUE
    -> WHERE (job_groups.batch_id = 1110000) AND (NOT deleted) AND (job_group_self_and_ancestors.ancestor_id = 0 AND job_group_self_and_ancestors.level = 1)
    -> ORDER BY job_group_id ASC;
+----+-------------------+--------------------------------------+------------+--------+--------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+-------------------------------------------------------+------+----------+---------------------------------+
| id | select_type       | table                                | partitions | type   | possible_keys                                                                                          | key                                            | key_len | ref                                                   | rows | filtered | Extra                           |
+----+-------------------+--------------------------------------+------------+--------+--------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+-------------------------------------------------------+------+----------+---------------------------------+
|  1 | PRIMARY           | batches                              | NULL       | const  | PRIMARY,batches_deleted                                                                                | PRIMARY                                        | 8       | const                                                 |    1 |   100.00 | Using temporary; Using filesort |
|  1 | PRIMARY           | job_group_self_and_ancestors         | NULL       | ref    | PRIMARY,job_group_self_and_ancestors_ancestor_id_level,job_group_self_and_ancestors_job_group_id_level | job_group_self_and_ancestors_ancestor_id_level | 16      | const,const,const                                     |    1 |   100.00 | Using index                     |
|  1 | PRIMARY           | job_groups                           | NULL       | eq_ref | PRIMARY,job_groups_state_callback,job_groups_time_created,job_groups_time_completed                    | PRIMARY                                        | 12      | const,batch.job_group_self_and_ancestors.job_group_id |    1 |   100.00 | Rematerialize (<derived2>)      |
|  1 | PRIMARY           | job_groups_n_jobs_in_complete_states | NULL       | const  | PRIMARY                                                                                                | PRIMARY                                        | 8       | const                                                 |    1 |   100.00 | Using where                     |
|  1 | PRIMARY           | job_groups_cancelled                 | NULL       | const  | PRIMARY                                                                                                | PRIMARY                                        | 8       | const                                                 |    1 |   100.00 | Using where                     |
|  1 | PRIMARY           | <derived2>                           | NULL       | ALL    | NULL                                                                                                   | NULL                                           | NULL    | NULL                                                  |    2 |   100.00 | Using where                     |
|  2 | DEPENDENT DERIVED | <derived3>                           | NULL       | ALL    | NULL                                                                                                   | NULL                                           | NULL    | NULL                                                  |    2 |   100.00 | Using filesort                  |
|  2 | DEPENDENT DERIVED | resources                            | NULL       | eq_ref | resource_id                                                                                            | resource_id                                    | 4       | usage_t.resource_id                                   |    1 |   100.00 | NULL                            |
|  3 | DEPENDENT DERIVED | aggregated_job_group_resources_v3    | NULL       | ref    | PRIMARY                                                                                                | PRIMARY                                        | 8       | batch.job_groups.batch_id                             |   21 |    10.00 | Using where; Using temporary    |
+----+-------------------+--------------------------------------+------------+--------+--------------------------------------------------------------------------------------------------------+------------------------------------------------+---------+-------------------------------------------------------+------+----------+---------------------------------+
9 rows in set, 3 warnings (0.00 sec)
```

Reply:
It's hard to make any determinations about the join order without actually having batches with nested job groups. Everything is finding exactly one record here.

Very personally, I find `FROM job_group_self_and_ancestors` a bit more intuitive because that's the table with multiple rows whereas we expect the other tables to have at most one matching row.